### ISSUE TITLE
[Android] onNotification callback not called when app in background or killed

### ISSUE THREAD

**Issue by JaimeFerBec** [fechanique/cordova-plugin-fcm#170]
### Problem
I had a really hard time finding out why, even though I was getting the notifications in the system's notification tray, all that happened when I clicked them was my app opening, but the onNotification callback was never being called.

### Solution
The solution is actually really simple: just add `"click_action":"FCM_PLUGIN_ACTIVITY"` to your message when you are sending one to an Android device from your backend server using Firebase.

This will trigger the creation of the FCMPluginActivity (I still do not fully understand how this is achieved just by setting the click_action on the message, though), which then sends the push notification's playload to your application via the onNotification callback.

### Documentation update request
I spent quite a lot of time reading this plugin's Android code, and debugging via adb trying to find out why the FCMPluginActivity was not created. I then found the solution in an answer to another one of this plugin's issues.
I am creating this issue only so that other people do not spend as much time as me trying to find the solution. Therefore, I think it would be a good idea to add a note about `click_action` in this plugin's documentation.

**Comment by tobika** [comment]
@JaimeFerBec thanks for the nice writeup :) you also get the onNotification when the app was killed (not in cache)? For me this is the only case that doesn't work still.

**Comment by JaimeFerBec** [comment]
@tobika Yes, in my case the onNotification callback gets called even if the app has been killed on Android. I'm still working it out for iOS though. When the app is killed on iOS, it won't do anything, and when the app is in background, the onNotification callback gets called only when I resume the app by tapping its icon on the start menu. The problem on iOS for me is that the notification never shows up on the system's notification tray. I'm still working on that issue.

**Comment by tobika** [comment]
Just to give a bit feedback and to confirm that everything works as it should :)

The event was sent after all (stupid me, didn't see it in the logs) but my app wasn't ready yet to take action yet (still loading). Thanks again for your help.

**Comment by wfdevelop** [comment]
@JaimeFerBec
I found out that for iOS it is important to use "priority" : "high". Only then the notification is shown, when app is killed or in background.

**Comment by JaimeFerBec** [comment]
@wfdevelop Notifications are now showing up in iOS. Thanks a lot!
I only have one more issue to solve now in iOS: when the app is killed, tapping the notification only wakes the app back up, but the onNotification callback is not called. I will investigate it further this evening to see what I can find out.

**Comment by StewartJingga** [comment]
@JaimeFerBec  hi, james, i was wondering if you could show me the payload for sending notification to iOS.

For some reasons, i am not receiving any notifications when app is on background.
I have set the priority to high and reinstalled the app. Still no luck.

From my understanding, the notification will be received regardless whether I have onNotification callback defined or not.  Or did i miss something?

**Comment by mamcgree** [comment]
Greetings ...

I am using your cordova-plugin-fcm in an ionic2 app on Android.  When notification is tapped the onNotification works when the app has been killed, pushed to the background and in the foreground.  The only time **it does not work** is after I do a this.platform.exitApp().  It looks like the the app is starting normally but no onNotification.  I think it might be because this.platform.exitApp() does not completely kill the app, because you can still see it in the task list.  Just wondering if you had any suggestions.

Cheers..
Mark..

**Comment by kalashri-webonise** [comment]
I have click_Action: 'FCM_PLUGIN_ACTIVITY' in the message which i am sending from server side.
 still onNotification() method is not getting executed on android.It is working fine on ios.
Below are cordova versions in my application.

`  "dependencies": {
    "cordova-android": "^6.2.3",
    "cordova-google-api-version": "0.0.1",
    "cordova-ios": "^4.5.0",
    "cordova-plugin-fcm": "git+https://github.com/ostownsville/cordova-plugin-fcm.git",
    "cordova-plugin-geolocation": "2.1.0",
    "cordova-plugin-inappbrowser": "1.2.0",
    "cordova-plugin-network-information": "1.2.1",
    "cordova-plugin-screen-orientation": "~2.0.2",
    "cordova-plugin-splashscreen": "3.2.2",
    "cordova-plugin-streaming-media": "^1.0.2",
    "cordova-plugin-whitelist": "1"
  }`

**Comment by fabiusss** [comment]
Hi, this is my solution [http://www.gestiscifacile.net/t50z](http://www.gestiscifacile.net/t50z).
It worked fine

**Comment by abcdurga** [comment]
Receiving push notifications when app is killed in android.
[Solved by this approach](https://github.com/fechanique/cordova-plugin-fcm/issues/502#issuecomment-396878069)

**Comment by indrabisen** [comment]
Thanks, it works for me.
